go/cmd/betabot/main.go

```diff
@@ -300,83 +334,124 @@ func (bot *Bot) handleUserMessageInteractionUpdated(ctx context.Context, _ *bert
 	}
 
 	var conv *Conversation
+	bot.storeMutex.RLock()
 	for i := range bot.store.Convs {
 		if bot.store.Convs[i].ConversationPublicKey == interaction.ConversationPublicKey {
-			conv = &bot.store.Convs[i]
+			conv = bot.store.Convs[i]
 		}
 	}
+	bot.storeMutex.RUnlock()
 	receivedMessage := payload.(*bertymessenger.AppMessage_UserMessage)
 	if conv != nil && conv.IsOneToOne {
-		switch {
-		case conv.Count == 0 && checkValidationMessage(receivedMessage.GetBody()):
-			conv.Count++
-			bot.saveStore()
-			time.Sleep(1 * time.Second)
+		unlock := bot.LockConversation(conv)
+		success, err := doSteps[conv.Step](ctx, conv, bot, receivedMessage, interaction, unlock)
+		if err != nil {
+			return err
+		}
+		if success {
+			return nil
+		}
+		// auto-reply to user's messages
+		answer := getRandomReply()
+		if err := bot.interactUserMessage(ctx, answer, interaction.ConversationPublicKey); err != nil {
+			return fmt.Errorf("interact user message failed: %w", err)
+		}
+	}
+	return nil
+}
+
+var doSteps map[uint]func(context.Context, *Conversation, *Bot, *bertymessenger.AppMessage_UserMessage, *bertymessenger.Interaction, func()) (bool, error)
```

Comment:
move this global and the init function just before using it (before line 347)

in general, avoid `func init()` and `globals`

Reply:
> in general, avoid `func init()` and `globals`

I know but I don't know where to create the function map.

Reply:
just before the call, before line 347